keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -131,75 +129,83 @@ Sidecar start up will follow these steps:
-1. The COSI controller copies the sidecar's secret to the `Bucket`'s namespace.
-1. The COSI controller adds a finalizer to the `BucketContent`, `Bucket`, and app secret.
-1. The COSI controller sets owner references on the `Bucket` and secret. Note: the `BucketContent` being cluster-scoped does not have an owner reference.
-1. The workflow ingests the Secret to begin operation.
-
-#### Revoke Access
-
-1. The user deletes the `Bucket`, which blocks until revoke operations complete.
-1. The COSI controller detects the update and deletes the `BucketContent`, which also blocks until backend revoke operations complete.
-1. The sidecar detects the `BucketContent` update and invokes the _RevokeAccess()_ rpc method.
-1. The driver removes access for the associated credentials.
-1. The sidecar removes the finalizer from and deletes its Secret.
-1. The sidecar sets the `BucketContent`'s status to indicate the revoke occurred.
-1. The COSI controller deletes the `BucketContent` and app secret.
-1. The COSI controller removes the `Bucket`, app secret, and `BucketContent` finalizers so they are garbage collected.
+### Distinguishing Bucket Case
+
+| BucketClassFields             | SecretRef: nil | SecretRef: non-nil |
+| ----------------------------- | -------------- | ------------------ |
+| **objectBucketName: non-nil** | Brownfield     | Undefined          |
+| **objectBucketName: nil**     | Greenfield     | Static             |
+
+#### Create Bucket (Greenfield)
+
+1. The user creates a `Bucket` in their namespace, with reference to a `BucketClass`.
+1. The Controller sees the new `Bucket` and applies a `finalizer` for orchestrated deletions.
+1. The Controller gets the `BucketClass` referenced by the `Bucket`.
+1. The Controller creates a `BucketContent` object with its `BucketClassName` set to the name of its `BucketClass` and a `finalizer`.
+1. The Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
+1. The Sidecar calls the CreateBucket() rpc, passing the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. The Sidecar creates a secret containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
+1. The Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”.*
+1. The Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
+1. The Controller copies the generated `secret` to the `Bucket` namespace with name defined in `Bucket.secretName`.  The `secret` is created with an `ownerRef` of the `Bucket`.
+1. The Controller “binds” the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
+1. The app `pod` ingests `secret` and runs.
+
+#### Grant Bucket Access (Brownfield)
```

Comment:
_Maybe_ greenfield and brownfield create can be combined. They are very similar and combining helps illustrate our goal of keeping brownfield and greenfield workflows similar.

Reply:
I agree, but I think we can leave this until after the meeting.